integration/setup/setup.go

```diff
@@ -228,29 +276,74 @@ func setupCollection(tb testing.TB, ctx context.Context, client *mongo.Client, o
 		span.End()
 	}
 
-	if len(opts.Providers) == 0 {
-		tb.Logf("Collection %s.%s wasn't created because no providers were set.", databaseName, collectionName)
-	} else {
-		require.True(tb, inserted, "all providers were not compatible")
+	return
+}
+
+// insertBenchmarkProvider inserts documents from specified BenchmarkProvider into collection.
+// It returns true if any document was inserted.
+//
+// The function calculates the checksum of all inserted documents and compare them with provider's hash.
+func insertBenchmarkProvider(tb testing.TB, ctx context.Context, collection *mongo.Collection, provider shareddata.BenchmarkProvider) (inserted bool) {
+	tb.Helper()
+
+	if provider == nil {
+		return
 	}
 
-	if ownCollection {
-		// delete collection and (possibly) database unless test failed
-		tb.Cleanup(func() {
-			if tb.Failed() {
-				tb.Logf("Keeping %s.%s for debugging.", databaseName, collectionName)
-				return
-			}
+	spanName := fmt.Sprintf("insertBenchmarkProvider/%s/%s", collection.Name(), provider.Hash())
+	provCtx, span := otel.Tracer("").Start(ctx, spanName)
+	region := trace.StartRegion(provCtx, spanName)
 
-			err := collection.Drop(ctx)
-			require.NoError(tb, err)
+	iter := provider.Docs()
+	hash := sha256.New()
 
-			if ownDatabase {
-				err = database.Drop(ctx)
-				require.NoError(tb, err)
+	for {
+		docs, err := iterator.ConsumeValuesN(iter, 10)
+		require.NoError(tb, err)
+
+		if len(docs) == 0 {
+			break
+		}
+
+		// convert []bson.D to []any, as mongodb requires.
+		var insertDocs []any = make([]any, len(docs))
+
+		for i, doc := range docs {
+			rawDoc := []byte(fmt.Sprintf("%x", doc))
+
+			// append literal document to previous checksum.
+			//
+			// We don't just write all the documents and Sum at the end, to not
+			// store all documents in memory.
+			currSum := hash.Sum(rawDoc)
+
+			// remove old checksum
+			hash.Reset()
+
+			// write literal document with previous checksum to calculate new checksum
+			if _, err = hash.Write(currSum); err != nil {
+				panic("Unexpected error: " + err.Error())
 			}
-		})
+
+			insertDocs[i] = doc
+		}
```

Comment:
Up to here is related to getting and verifying providers, shouldn't it be in shareddata rather than setup?

Reply:
I wasn't 100% sure where to place the checksum but I've noticed that we already do some validation for the normal `Provider` interface inside the setup function.

The problem with extracting the validation logic to the BenchmarkValues, is the fact that we allow developer to introduce different structs that implement `BenchmarkProvider`, which would omit the checksum checking. 

I believe that validating the correct order of benchmark documents is one of the essential parts of the benchmark, and it seems to me that it should be enforced on every struct that implements BenchmarkProvider interface.


The other thing is that implementation would be slightly more complicated, as we would need to wrap iterator of BenchmarkValues into another iterator, that would calculate the checksum on-go. Also, BenchmarkValues would need to do some error handling because of that.

Reply:
I see it now. We read the values first time here so we don't want to do it a provider level